core/src/test/scala/unit/kafka/server/TxnOffsetCommitRequestTest.scala

```diff
@@ -0,0 +1,149 @@
+    val groupId = "group"
+
+    // Creates the __consumer_offsets topics because it won't be created automatically
+    // in this test because it does not use FindCoordinator API.
+    createOffsetsTopic()
+
+    // Join the consumer group. Note that we don't heartbeat here so we must use
+    // a session long enough for the duration of the test.
+    joinConsumerGroup(groupId, useNewProtocol)
+
+    var consumer: Consumer[Bytes, Bytes] = null
+    var producer: Producer[String, String] = null
+    try {
+      createTopic(topic, 1)
+
+      producer = cluster.producer(Collections.singletonMap(ProducerConfig.TRANSACTIONAL_ID_CONFIG, transactionalId))
+      producer.initTransactions()
+      producer.beginTransaction()
+      addOffsetsToTxn(groupId, 0, 0.toShort, transactionalId, ApiKeys.ADD_OFFSETS_TO_TXN.latestVersion())
+
+      val consumerConfigs = new util.HashMap[String, Object]()
+      consumerConfigs.put(ConsumerConfig.EXCLUDE_INTERNAL_TOPICS_CONFIG, "false")
+      consumer = cluster.consumer(consumerConfigs)
+      consumer.assign(Collections.singletonList(new TopicPartition(Topic.GROUP_METADATA_TOPIC_NAME, 0)))
+
+      // Verify that the TXN_OFFSET_COMMIT request is processed correctly when member id is UNKNOWN_MEMBER_ID
+      // and generation id is UNKNOWN_GENERATION_ID under all api versions
+      for (version <- 0 to ApiKeys.TXN_OFFSET_COMMIT.latestVersion(isUnstableApiEnabled)) {
+        commitTxnOffset(
+          groupId = groupId,
+          memberId = JoinGroupRequest.UNKNOWN_MEMBER_ID,
+          generationId = JoinGroupRequest.UNKNOWN_GENERATION_ID,
+          producerId = 0,
+          producerEpoch = 0.toShort,
+          transactionalId = transactionalId,
+          topic = topic,
+          partition = partition,
+          offset = 100 + version,
+          expectedError = Errors.NONE,
+          version = version.toShort,
+        )
+
+        TestUtils.waitUntilTrue(() => {
+          consumer.poll(Duration.ofSeconds(5)).asScala
+            .filter(record => record.key() != null && record.value() != null)
+            .map(record => new GroupCoordinatorRecordSerde()
+              .deserialize(ByteBuffer.wrap(record.key().get()), ByteBuffer.wrap(record.value().get())))
+            .filter(coordinatorRecord =>
+              coordinatorRecord.key().message().isInstanceOf[OffsetCommitKey] &&
+              coordinatorRecord.value().message().isInstanceOf[OffsetCommitValue])
+            .exists(coordinatorRecord => {
+              val offsetCommitKey = coordinatorRecord.key().message().asInstanceOf[OffsetCommitKey]
+              val offsetCommitValue = coordinatorRecord.value().message().asInstanceOf[OffsetCommitValue]
+              offsetCommitKey.group() == groupId &&
+                offsetCommitKey.topic() == topic &&
+                offsetCommitKey.partition == partition &&
+                offsetCommitValue.offset() == 100 + version
+            })
+        }, "Txn offset commit not found")
```

Comment:
This is a bit extreme. I would rather prefer to use `fetchOffset` to validate the committed offsets.

Reply:
Thanks for the review, I have addressed all comments in the latest commits. Also loop the integration test 247 times and all passed.
```sh
I=0; while ./gradlew cleanTest core:test --tests kafka.server.TxnOffsetCommitRequestTest --rerun --fail-fast; do (( I=$I+1 )); echo "Completed run: $I"; sleep 1; done
```